pennylane/optimize/shot_adaptive.py

```diff
@@ -0,0 +1,464 @@
+            number of shots recommended for each gradient component changes.
+        b (float): Regularization bias. The bias should be kept small, but non-zero.
+        weighted_random_sampling (bool): Whether to use the weighted random sampling algorithm
+            to multinomially distribute the shot budget across terms in the Hamiltonian expectation value.
+            Only takes effect if the objective function provided is an instance of :class:`~.ExpvalCost`.
+        stepsize (float): The learning rate :math:`\eta`. The learning rate *must* be such
+            that :math:`\eta < 2/L = 2/\sum_i|c_i|`, where:
+
+            * :math:`L \leq \sum_i|c_i|` is the bound on the `Lipschitz constant
+              <https://en.wikipedia.org/wiki/Lipschitz_continuity>`__ of the variational quantum
+              algorithm objective function, and
+
+            * :math:`c_i` are the coefficients of the Hamiltonian used in the objective function.
+
+    The shot adaptive optimizer is based on the iCANS1 optimizer by
+    `Kübler et al. (2020) <https://quantum-journal.org/papers/q-2020-05-11-263/>`__, and works
+    as follows:
+
+    1. The initial step of the optimizer is performed with some specified minimum
+       number of shots, :math:`s_{min}`, for all partial derivatives.
+
+    2. The parameter-shift rule is then used to estimate the gradient :math:`g_i`
+       for each parameter :math:`\theta_i`, parameters, as well as the *variances*
+       :math:`v_i` of the estimated gradients.
+
+    3. Gradient descent is performed for each parameter :math:`\theta_i`, using
+       the pre-defined learning rate :math:`\eta` and the gradient information :math:`g_i`:
+       :math:`\theta_i \rightarrow \theta_i - \eta g_i`.
+
+    4. The improvement in the expected gain per shot, for a specific parameter value,
+       is then calculated via
+
+       .. math::
+           \gamma_i = \frac{1}{s_i} \left[ \left(\eta - \frac{1}{2} L\eta^2\right)
+                       g_i^2 - \frac{L\eta^2}{2s_i}v_i \right],
+
+       where:
+
+       * :math:`L \leq \sum_i|c_i|` is the bound on the `Lipschitz constant
+         <https://en.wikipedia.org/wiki/Lipschitz_continuity>`__ of the variational quantum algorithm objective function,
+
+       * :math:`c_i` are the coefficients of the Hamiltonian, and
+
+       * :math:`\eta` is the learning rate, and *must* be bound such that :math:`\eta < 2/L`
+         for the above expression to hold.
+
+    5. Finally, the new values of :math:`s_i` (shots for partial derivative of parameter
+       :math:`\theta_i`) is given by:
+
+       .. math::
+
+           s_i = \frac{2L\eta}{2-L\eta}\left(\frac{v_i}{g_i^2}\right)\propto
+                 \frac{v_i}{g_i^2}.
+
+    In addition to the above, to counteract the presence of noise in the system, a
+    running average of :math:`g_i` and :math:`s_i` (:math:`\chi_i` and :math:`\xi_i` respectively)
+    are used when computing :math:`\gamma_i` and :math:`s_i`.

```

Comment:
iCANS and Rosalin are based on top of Adam.  Is this implementation also based on Adam?  Might be good to mention if so.

Reply:
Good point, I'll need to think how to word this...